Error with truncate table + index

**BUG REPORT**

1. Please describe the issue you observed:
Error while truncating a table with a secondary index and existing data

- What did you do?
Table truncate

- What did you expect to see?
The table truncated

- What did you see instead?
```
ERROR:java.sql.SQLException: herddb.core.HerdDBInternalException: herddb.model.StatementExecutionException: internal data error: herddb.storage.DataStorageManagerException: java.io.IOException: herddb.storage.DataStorageManagerException: java.nio.file.NoSuchFileException: /home/root/*****/dbdata/data/7505a889a4c0432e870edf97946f1c2c.tablespace/543624a9-887f-445b-8a6f-698481767c19.index/2.page
        at herddb.core.TableManager.lambda$executeStatementAsync$1(TableManager.java:581)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
        at herddb.core.TableManager.executeStatementAsync(TableManager.java:576)
        at herddb.core.TableSpaceManager.executeTableAwareStatement(TableSpaceManager.java:1226)
        at herddb.core.TableSpaceManager.executeStatementAsyncInternal(TableSpaceManager.java:1122)
        at herddb.core.TableSpaceManager.executeStatementAsync(TableSpaceManager.java:1092)
        at herddb.core.DBManager.executeStatementAsync(DBManager.java:648)
        at herddb.core.DBManager.executePlanAsync(DBManager.java:705)
        at herddb.server.ServerSideConnectionPeer.handleExecuteStatement(ServerSideConnectionPeer.java:721)
        at herddb.server.ServerSideConnectionPeer.requestReceived(ServerSideConnectionPeer.java:138)
        at herddb.network.netty.NettyChannel.lambda$handlePduRequest$0(NettyChannel.java:96)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: herddb.model.StatementExecutionException: internal data error: herddb.storage.DataStorageManagerException: java.io.IOException: herddb.storage.DataStorageManagerException: java.nio.file.NoSuchFileException: /home/root/****/dbdata/data/7505a889a4c0432e870edf97946f1c2c.tablespace/543624a9-887f-445b-8a6f-698481767c19.index/2.page
        ... 19 more
Caused by: herddb.storage.DataStorageManagerException: java.io.IOException: herddb.storage.DataStorageManagerException: java.nio.file.NoSuchFileException: /home/root/****/dbdata/data/7505a889a4c0432e870edf97946f1c2c.tablespace/543624a9-887f-445b-8a6f-698481767c19.index/2.page
        at herddb.index.brin.BRINIndexManager.checkpoint(BRINIndexManager.java:369)
        at herddb.core.TableManager.checkpoint(TableManager.java:2756)
        at herddb.core.TableManager.checkpoint(TableManager.java:2180)
        at herddb.core.TableManager.flush(TableManager.java:1992)
        at herddb.core.TableManager.lambda$executeStatementAsync$1(TableManager.java:579)
        ... 18 more
Caused by: java.io.IOException: herddb.storage.DataStorageManagerException: java.nio.file.NoSuchFileException: /home/root/****/dbdata/data/7505a889a4c0432e870edf97946f1c2c.tablespace/543624a9-887f-445b-8a6f-698481767c19.index/2.page
        at herddb.index.brin.BRINIndexManager$IndexDataStorageImpl.createDataPage(BRINIndexManager.java:474)
        at herddb.index.brin.BlockRangeIndex$Block.checkpointNoLock(BlockRangeIndex.java:665)
        at herddb.index.brin.BlockRangeIndex$Block.checkpoint(BlockRangeIndex.java:679)
        at herddb.index.brin.BlockRangeIndex$Block.access$800(BlockRangeIndex.java:220)
        at herddb.index.brin.BlockRangeIndex.merge(BlockRangeIndex.java:760)
        at herddb.index.brin.BlockRangeIndex.checkpoint(BlockRangeIndex.java:980)
        at herddb.index.brin.BRINIndexManager.checkpoint(BRINIndexManager.java:310)
        ... 22 more
Caused by: herddb.storage.DataStorageManagerException: java.nio.file.NoSuchFileException: /home/root/****/dbdata/data/7505a889a4c0432e870edf97946f1c2c.tablespace/543624a9-887f-445b-8a6f-698481767c19.index/2.page
        at herddb.file.FileDataStorageManager.writeIndexPage(FileDataStorageManager.java:947)
        at herddb.index.brin.BRINIndexManager$IndexDataStorageImpl.createDataPage(BRINIndexManager.java:470)
        ... 28 more
Caused by: java.nio.file.NoSuchFileException: /home/root/****/dbdata/data/7505a889a4c0432e870edf97946f1c2c.tablespace/543624a9-887f-445b-8a6f-698481767c19.index/2.page
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:178)
        at java.base/java.nio.channels.FileChannel.open(FileChannel.java:292)
        at java.base/java.nio.channels.FileChannel.open(FileChannel.java:345)
        at herddb.utils.ManagedFile.open(ManagedFile.java:79)
        at herddb.file.FileDataStorageManager.writeIndexPage(FileDataStorageManager.java:938)
        ... 29 more
```

--- a/herddb-core/src/main/java/herddb/core/AbstractIndexManager.java
+++ b/herddb-core/src/main/java/herddb/core/AbstractIndexManager.java
@@ -182,6 +182,20 @@ public abstract class AbstractIndexManager implements AutoCloseable {
         dataStorageManager.dropIndex(tableSpaceUUID, index.uuid);
     }
 
+    /**
+     * Truncate the index from persist storage
+     * <p>
+     * Differs from {@link #dropIndexData()} because it leaves in place every support structure needed by
+     * index runtime (for example if index is persisted into a directory it doesn't delete the directory
+     * itself). The index will continue to work but without current data that will be erased
+     * </p>
+     *
+     * @throws DataStorageManagerException
+     */
+    public void truncateIndexData() throws DataStorageManagerException {
+        dataStorageManager.truncateIndex(tableSpaceUUID, index.uuid);
+    }
+
     final void onTransactionCommit(Transaction transaction, boolean recovery) throws DataStorageManagerException {
         if (createdInTransaction > 0) {
             if (transaction.transactionId != createdInTransaction) {

--- a/herddb-core/src/main/java/herddb/file/FileDataStorageManager.java
+++ b/herddb-core/src/main/java/herddb/file/FileDataStorageManager.java
@@ -40,6 +40,8 @@ import herddb.storage.IndexStatus;
 import herddb.storage.TableStatus;
 import herddb.utils.ByteArrayCursor;
 import herddb.utils.Bytes;
+import herddb.utils.CleanDirectoryFileVisitor;
+import herddb.utils.DeleteFileVisitor;
 import herddb.utils.ExtendedDataInputStream;
 import herddb.utils.ExtendedDataOutputStream;
 import herddb.utils.FileUtils;
@@ -59,14 +61,12 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.file.DirectoryStream;
-import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
 import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
 import java.nio.file.SimpleFileVisitor;
 import java.nio.file.StandardCopyOption;
 import java.nio.file.StandardOpenOption;
-import java.nio.file.attribute.BasicFileAttributes;
 import java.nio.file.attribute.FileTime;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -1281,6 +1281,17 @@ public class FileDataStorageManager extends DataStorageManager {
     }
 
     @Override
+    public void truncateIndex(String tablespace, String name) throws DataStorageManagerException {
+        Path tableDir = getIndexDirectory(tablespace, name);
+        LOGGER.log(Level.INFO, "truncateIndex {0}.{1} in {2}", new Object[]{tablespace, name, tableDir});
+        try {
+            cleanDirectory(tableDir);
+        } catch (IOException ex) {
+            throw new DataStorageManagerException(ex);
+        }
+    }
+
+    @Override
     public void dropIndex(String tablespace, String name) throws DataStorageManagerException {
         Path tableDir = getIndexDirectory(tablespace, name);
         LOGGER.log(Level.INFO, "dropIndex {0}.{1} in {2}", new Object[]{tablespace, name, tableDir});
@@ -1339,27 +1350,19 @@ public class FileDataStorageManager extends DataStorageManager {
         }
     }
 
-    private static class FileDeleter extends SimpleFileVisitor<Path> {
-
-        @Override
-        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-//            println("delete file " + file.toAbsolutePath());
-            Files.delete(file);
-            return FileVisitResult.CONTINUE;
-        }
-
-        @Override
-        public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
-//            println("delete directory " + dir);
-            Files.delete(dir);
-            return FileVisitResult.CONTINUE;
-        }
+    public static void deleteDirectory(Path f) throws IOException {
+        deleteDirectoryContent(f, false);
     }
 
-    public static void deleteDirectory(Path f) throws IOException {
+    public static void cleanDirectory(Path f) throws IOException {
+        deleteDirectoryContent(f, true);
+    }
+
+    private static void deleteDirectoryContent(Path f, boolean keepDirectory) throws IOException {
         if (Files.isDirectory(f)) {
-            Files.walkFileTree(f, new FileDeleter());
-            Files.deleteIfExists(f);
+            final SimpleFileVisitor<Path> visitor =
+                    keepDirectory ? new CleanDirectoryFileVisitor() : DeleteFileVisitor.INSTANCE;
+            Files.walkFileTree(f, visitor);
         } else if (Files.isRegularFile(f)) {
             throw new IOException("name " + f.toAbsolutePath() + " is not a directory");
         }
@@ -1367,10 +1370,7 @@ public class FileDataStorageManager extends DataStorageManager {
 
     @Override
     public KeyToPageIndex createKeyToPageMap(String tablespace, String name, MemoryManager memoryManager) throws DataStorageManagerException {
-
         return new BLinkKeyToPageIndex(tablespace, name, memoryManager, this);
-
-//        return new ConcurrentMapKeyToPageIndex(new ConcurrentHashMap<>());
     }
 
     @Override

--- a/herddb-core/src/main/java/herddb/index/brin/BRINIndexManager.java
+++ b/herddb-core/src/main/java/herddb/index/brin/BRINIndexManager.java
@@ -479,7 +479,7 @@ public class BRINIndexManager extends AbstractIndexManager {
     @Override
     public void truncate() throws DataStorageManagerException {
         this.data.reset();
-        dropIndexData();
+        truncateIndexData();
     }
 
     @Override

--- a/herddb-core/src/main/java/herddb/mem/MemoryDataStorageManager.java
+++ b/herddb-core/src/main/java/herddb/mem/MemoryDataStorageManager.java
@@ -542,6 +542,11 @@ public class MemoryDataStorageManager extends DataStorageManager {
     }
 
     @Override
+    public void truncateIndex(String tableSpaceUUID, String name) throws DataStorageManagerException {
+        dropIndex(tableSpaceUUID, name);
+    }
+
+    @Override
     public void dropIndex(String tablespace, String name) throws DataStorageManagerException {
         List<Index> indexes = indexesByTablespace.get(tablespace);
         if (indexes != null) {

--- a/herddb-core/src/main/java/herddb/storage/DataStorageManager.java
+++ b/herddb-core/src/main/java/herddb/storage/DataStorageManager.java
@@ -247,6 +247,8 @@ public abstract class DataStorageManager implements AutoCloseable {
     public abstract void cleanupAfterBoot(String tablespace, String name, Set<Long> activePagesAtBoot)
             throws DataStorageManagerException;
 
+    public abstract void truncateIndex(String tableSpaceUUID, String name) throws DataStorageManagerException;
+
     public abstract void dropIndex(String tableSpaceUUID, String name) throws DataStorageManagerException;
 
     /* Map[tablespace_uuid,Map[pageid,pincounts] */

--- a/herddb-core/src/test/java/herddb/core/TruncateTableSQLTest.java
+++ b/herddb-core/src/test/java/herddb/core/TruncateTableSQLTest.java
@@ -24,6 +24,7 @@ import static herddb.core.TestUtils.execute;
 import static herddb.core.TestUtils.scan;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.fail;
+import herddb.file.FileDataStorageManager;
 import herddb.mem.MemoryCommitLogManager;
 import herddb.mem.MemoryDataStorageManager;
 import herddb.mem.MemoryMetadataStorageManager;
@@ -33,8 +34,11 @@ import herddb.model.StatementExecutionException;
 import herddb.model.TableDoesNotExistException;
 import herddb.model.TransactionContext;
 import herddb.model.commands.CreateTableSpaceStatement;
+import java.nio.file.Path;
 import java.util.Collections;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
 
 /**
  * Tests on table creation
@@ -42,6 +46,9 @@ import org.junit.Test;
  * @author enrico.olivelli
  */
 public class TruncateTableSQLTest {
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
 
     @Test
     public void truncateTableNoTransaction() throws Exception {
@@ -112,6 +119,37 @@ public class TruncateTableSQLTest {
             try (DataScanner scan = scan(manager, "SELECT * FROM tblspace1.tsql ", Collections.emptyList())) {
                 assertEquals(0, scan.consume().size());
             }
+
+        }
+
+    }
+
+    @Test
+    public void truncateTableFileDataStore() throws Exception {
+        String nodeId = "localhost";
+        Path dataPath = folder.newFolder("data").toPath();
+        try (DBManager manager = new DBManager("localhost", new MemoryMetadataStorageManager(),
+                new FileDataStorageManager(dataPath), new MemoryCommitLogManager(), null, null)) {
+            manager.start();
+            CreateTableSpaceStatement st1 =
+                    new CreateTableSpaceStatement("tblspace1", Collections.singleton(nodeId), nodeId, 1, 0, 0);
+            manager.executeStatement(
+                    st1, StatementEvaluationContext.DEFAULT_EVALUATION_CONTEXT(), TransactionContext.NO_TRANSACTION);
+            manager.waitForTablespace("tblspace1", 10000);
+
+            execute(manager, "CREATE TABLE tblspace1.tsql (k1 string primary key,n1 int not nul,s1 string)", Collections.emptyList());
+            execute(manager, "CREATE INDEX test1 ON tblspace1.tsql (n1)", Collections.emptyList());
+            execute(manager, "INSERT INTO tblspace1.tsql (k1,n1) values('a',1)", Collections.emptyList());
+
+            try (DataScanner scan = scan(manager, "SELECT * FROM tblspace1.tsql ", Collections.emptyList())) {
+                assertEquals(1, scan.consume().size());
+            } catch (TableDoesNotExistException ok) {}
+
+            execute(manager, "TRUNCATE TABLE tblspace1.tsql", Collections.emptyList());
+
+            try (DataScanner scan = scan(manager, "SELECT * FROM tblspace1.tsql ", Collections.emptyList())) {
+                assertEquals(0, scan.consume().size());
+            } catch (TableDoesNotExistException ok) {}
 
         }
 

--- a/herddb-utils/src/main/java/herddb/utils/CleanDirectoryFileVisitor.java
+++ b/herddb-utils/src/main/java/herddb/utils/CleanDirectoryFileVisitor.java
@@ -1,0 +1,68 @@
+/*
+ Licensed to Diennea S.r.l. under one
+ or more contributor license agreements. See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership. Diennea S.r.l. licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+
+ */
+
+package herddb.utils;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.FileVisitor;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+
+/**
+ * A {@link FileVisitor} which delete every file o directory found (excluding starting directory).
+ * <p>
+ * If the visitor is executed on a path denoting a simple file it will just delete the file and exit.
+ * </p>
+ *
+ * @author diego.salvi
+ */
+public final class CleanDirectoryFileVisitor extends SimpleFileVisitor<Path> {
+
+    private int level;
+
+    public CleanDirectoryFileVisitor() {
+        level = 0;
+    }
+
+    @Override
+    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        Files.delete(file);
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+        // Delete only if is the first directory (level 0) */
+        if (--level > 0) {
+            Files.delete(dir);
+        }
+
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+        ++level;
+        return FileVisitResult.CONTINUE;
+    }
+}

--- a/herddb-utils/src/main/java/herddb/utils/DeleteFileVisitor.java
+++ b/herddb-utils/src/main/java/herddb/utils/DeleteFileVisitor.java
@@ -1,0 +1,53 @@
+/*
+ Licensed to Diennea S.r.l. under one
+ or more contributor license agreements. See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership. Diennea S.r.l. licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+
+ */
+
+package herddb.utils;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.FileVisitor;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+
+/**
+ * A {@link FileVisitor} which delete every file o directory found (including starting directory).
+ *
+ * @author diego.salvi
+ */
+public class DeleteFileVisitor extends SimpleFileVisitor<Path> {
+
+    public static final SimpleFileVisitor<Path> INSTANCE = new DeleteFileVisitor();
+
+    private DeleteFileVisitor() {}
+
+    @Override
+    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        Files.delete(file);
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+        Files.delete(dir);
+        return FileVisitResult.CONTINUE;
+    }
+}

--- a/herddb-utils/src/main/java/herddb/utils/FileUtils.java
+++ b/herddb-utils/src/main/java/herddb/utils/FileUtils.java
@@ -25,12 +25,9 @@ import java.io.IOException;
 import java.io.RandomAccessFile;
 import java.nio.ByteBuffer;
 import java.nio.channels.SeekableByteChannel;
-import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.SimpleFileVisitor;
 import java.nio.file.StandardOpenOption;
-import java.nio.file.attribute.BasicFileAttributes;
 
 /**
  * Utilities
@@ -46,20 +43,7 @@ public class FileUtils {
         if (!Files.isDirectory(directory)) {
             return;
         }
-        Files.walkFileTree(directory, new SimpleFileVisitor<Path>() {
-            @Override
-            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-                Files.delete(file);
-                return FileVisitResult.CONTINUE;
-            }
-
-            @Override
-            public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
-                Files.delete(dir);
-                return FileVisitResult.CONTINUE;
-            }
-
-        });
+        Files.walkFileTree(directory, DeleteFileVisitor.INSTANCE);
     }
 
     public static byte[] fastReadFile(Path f) throws IOException {

